runtime/lua/vim/fcnotify.lua

```diff
@@ -0,0 +1,172 @@
+-- Create a file watcher, each watcher is identified by the name of the file that it
+-- watches.
+
+local uv = vim.loop
+local Watcher = {}
+Watcher.__index = Watcher
+local w = {}
+local WatcherList = {}
+local check_handle = nil
+
+--- Callback for the check handle, checks if there are pending notifications
+--- for any watcher, and handles them as per the value of the `fcnotify`
+--- option.
+local function check_notifications()
+  for _, watcher in pairs(WatcherList) do
+    local status, option = pcall(vim.api.nvim_buf_get_option, watcher.bufnr, 'filechangenotify')
+    if not status then
+      option = vim.api.nvim_get_option('filechangenotify')
+    end
+    if option == 'off' or option:find('watcher') == nil then
+      watcher.pending_notifs = false
+    elseif watcher.pending_notifs then
+      vim.api.nvim_command('checktime '..watcher.bufnr)
+      watcher.pending_notifs = false
+    end
+  end
+end
+
+--- Checks if a buffer should have a watcher attached to it.
+--- Ignores all the buffers that aren't listed, or have a buf id
+--- less than 0.
+--- Ignores the [NO NAME] buffer.
+---
+--@param fname: The validity of this buffer will be checked.
+local function valid_buf(fname)
+  if fname == '' then
+    return false
+  end
+
+  local bufnr = vim.api.nvim_call_function('bufnr', {fname})
+
+  if bufnr < 0 then
+    return false
+  end
+
+  local buflisted = vim.api.nvim_buf_get_option(bufnr, 'buflisted')
+  local buftype = vim.api.nvim_buf_get_option(bufnr, 'buftype')
+
+  if not buflisted or buftype == 'nofile' or buftype == 'quickfix' then
+    return false
+  end
+  return true
+end
+
+--- Creates and initializes a new watcher object with the given filename.
+---
+--@param fname: (required, string) The path that the watcher should watch.
+function Watcher:new(fname)
+  vim.validate{fname = {fname, 'string', false}}
+  -- get full path name for the file
+  local ffname = vim.api.nvim_call_function('fnamemodify', {fname, ':p'})
+  w = {bufnr = vim.api.nvim_call_function('bufnr', {fname}),
+       fname = fname, ffname = ffname, handle = nil,
+       paused = false, pending_notifs = false}
+  setmetatable(w, self)
+  return w
+end
+
+--- Starts the watcher
+---
+--@param self: (required, table) The watcher object on which start was called.
+function Watcher:start()
+  self.handle = uv.new_fs_event()
+  self.handle:start(self.ffname, {}, function(...)
+    self:on_change(...)
+  end)
+end
+
+--- Stops the watcher and closes the handle.
+---
+--@param self: (required, table) The watcher object on which stop was called.
+function Watcher:stop()
+  self.handle:stop()
+
+  -- close the handle altogether, for windows.
+  if self.handle:is_closing() then
+    return
+  end
+  self.handle:close()
```

Comment:
why the check for is_closing before close is called ? does't `close` check for `is_closing` ?

Reply:
No it doesn't. Close simply requests for a handle to be closed.